ps2.h

```diff
@@ -382,45 +388,13 @@ class PS2KeyboardPort : public PS2Port<clkPin, datPin, size>
     /// @brief Converts a PS/2 Set 2 extended scan code to a IBM System/2 key number
     /// @param scancode The second byte of an extended scan code (after 0xe0)
     uint8_t ps2ext_to_keycode(uint8_t scancode) {
-      switch (scancode) {
-        case 0x11:  // Right Alt
-          return 62;
-        case 0x14:  // Right Ctrl
-          return 64;
-        case 0x1f:  // Left GUI
-          return 59;
-        case 0x27:  // Right GUI
-          return 63;
-        case 0x2f:  // Menu key
-          return 65;
-        case 0x69:  // End
-          return 81;
-        case 0x70:  // Insert
-          return 75;
-        case 0x71:  // Delete
-          return 76;
-        case 0x6b:  // Left arrow
-          return 79;
-        case 0x6c:  // Home
-          return 80;
-        case 0x75:  // Up arrow
-          return 83;
-        case 0x72:  // Down arrow
-          return 84;
-        case 0x7d:  // Page up
-          return 85;
-        case 0x7a:  // Page down
-          return 86;
-        case 0x74:  // Right arrow
-          return 89;
-        case 0x4a:  // KP divide
-          return 95;
-        case 0x5a:  // KP enter
-          return 108;
-        case 0x7c:  // KP PrtScr
-          return 124;
-        case 0x15:  // Pause/Break
-          return 126;
+      // todo replace with binary search for better speed?
```

Comment:
Not sure that we need better speed, but if we do, we also have 512 bytes of EEPROM unused and could put a lookup table there.

Reply:
I do not know how fast this code needs to be. Is anyone able to test this code somehow, to ensure that a linear search is fast enough? If there are many elements to consider, binary search may be a little faster, but will likely cause the code to become bigger. Sure, a 256 byte lookup table in eeprom would make it fast, but that would add both complexity and risk of it getting corrupt somehow, if we do expose an eeprom API

Reply:
According to specs, PS/2 minimum clock cycle is 60 us (16.67 KHz). From receiving the 11th bit, I think we are not guaranteed more time than that. 

In practice, the extended scan code table in your PR code is only searched when the firmware expects to receive the last byte of an extended scan code. If you hold down a key, scan codes are sent with significant delay (in the range of tens of seconds), and the firmware has all the time to process a scan code. Maybe if you press/release multiple keys at the same time you can get a new scan code without any delay in between. But that is virtually impossible to test just by typing on a keyboard.

I'm not skilled enough to fluently read the assembly produced by the C++ compiler. I think the code that loops over the extended scan code table ends with a comparison here, 19 being the length of the table:

```
3b2:	83 31       	cpi	r24, 0x13	; 19
3b4:	91 05       	cpc	r25, r1
3b6:	51 f7         	brne .-44     	; 0x38c <__stack+0x12d>
```

That jumps back to 0x38c, which I think is the start of the loop. Between the start of the loop and the end of the loop, about 10 instructions are executed. Most of them are 1 cycle, but there are also an rjmp and a brne taking 2 cycles and an lpm instruction taking 3. Without calculating it exactly, it's about 15 cycles per loop, about 285 cycles total for the entire search.

At 16 MHz we have 960 cycles to spare if to keep within 60 us, so 285 cycles are significant. It would of coarse be great to find a quicker way.

On the other hand, even if I run the SMC at 8 MHz, I haven't seen any failed key strokes. Removing the Arduino library functions doesn't just save space, it also speeds up execution of the firmware.

Reply:
Thanks for the input, I will try to find the loop and count worst case delay. If we are close to the limit, doing a binary search may improve the speed. If we are really close to the limit, a lookup table would be even faster, but that would for sure increase code size.

Reply:
(avr-gcc usually produces quite efficient code if you know how to use it, but its assembly listing is hard to decode compared to other compilers....)

Reply:
Worst case with binary search over 19 elements would be like 5 iterations, right? A little more code to execute, but probably faster than the current PR code.

The lookup table might not need to be 256 bytes. The lowest value is 0x11 and the highest 0x7d. A lookup table of 109 elements would cover the range.

Reply:
Just of curiosity, I tried to implement a 109 byte lookup table for extended scan codes. This grew the firmware from 6,474 bytes to 6,520 bytes, a difference of 46 bytes.

```
const uint8_t PS2_EXT_SCANCODES[] PROGMEM = {
  62, 0, 0, 64, 126, 0, 0, 0, 0, 0, 0, 
  0, 0, 0, 59, 0, 0, 0, 0, 0, 0, 0, 63, 
  0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
  0, 0, 0, 0, 0, 0, 0, 0, 0, 95, 0, 0, 0, 
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
  0, 108, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
  0, 0, 0, 0, 0, 81, 0, 79, 80, 0, 0, 0,
  75, 76, 84, 0, 89, 83, 0, 0, 0, 0, 
  86, 0, 124, 85
};
```

Reply:
Analyzing the build output, this is actually exactly what GCC does already before my changes (making a LUT from 0x11 to 0x7d). If you check the hex file, this lookup table is at the end of flash. This it copied into ram on startup, and is used as a quite fast lookup table in ram. Worst case, using this LUT, is 9 clock cycles. The disadvantage of GCC's solution is that this occupies a 108 byte lookup table in both flash and ram.

Reply:
The execution time of the current linear search algorithm is 16 clock cycles if the first one match, and 14 clock cycle for all following tests. For 19 elements, worst case seem to be 268 clock cycles. At 16 MHz, this is 15.75 us.

A quick asm optimized inner loop can be done in 8 clock cycles for lpm, or 7 clock cycles if we copy 19 bytes to ram. Thus, the worst case loop is 144 or 126 cycles (with or without lpm) plus overhead, almost half the execution time of the current C code.

Reply:
A binary search inner loop can be done in 9 or 10 cycles pr loop. Then it will always be 5 loops if there are up to 32 elements. Requiring 32 bytes in ram, and 32+19 bytes LUT in flash, and code. (inner loop is 22 bytes)
```
next:
ld r16,Z ; mid
cpi r16,r17
breq match
brlo lower
;not lower
add r30,r18
lsr r18
brne next
rjmp not_found
lower:
sub r30,r18
lsr r18
brne next
not_found:
```

Reply:
Another option can be to use LUT in ram, as done originally (consuming 108 byte ram), but using some custom code to initialize it, to reduce flash usage. I guess it should be doable in less than 108 bytes, which the original implementation uses.

Reply:
Now that I know that the original implementation is a lookup table already, I suggest to keep it that way. Changing from LUT to linear search saves only 36 bytes of flash, and 110 bytes of ram, but speed is much slower, and in this case, speed seem to matter.

If the ram is needed, a future improvement may implement this LUT explicitly, and use LPM instead of reading from a copy in ram. As I guess we can afford 1 extra clock cycle. (LPM (from flash) is 3 clock cycles, while LD (from ram) is 2 clock cycles)

Saving 36 bytes of flash is not important at all, for now. Thus, I will revert this to the original implementation.

Reply:
I have now reverted to the original implementation, which causes a look-up table to be generated and placed in ram. Resolving this conversation.